### PR TITLE
[checkbox-ce-oem] Make serial test (console and transmit data) automated (New)

### PR DESCRIPTION
**Pull request by LiaoU3**
## Description
I created automated serial test cases including checking serial console and transmitting data (including send and receive). These test cases could replace the old manual and automated serial test cases.

#### serial console
Loopback the DUT with usb-to-serial cable
#### transmit data
Connect serial ports with a serial echo server (e.g. another serial port on device)

## Resolved issues
With these new test cases, we no longer need manual test for serial ports

## Documentation
Will make it up in the future

## Tests
Serial console
 - https://certification.canonical.com/hardware/202307-31906/submission/360200/

Transmitting data
 - https://certification.canonical.com/hardware/202401-33430/submission/359950/
 - https://certification.canonical.com/hardware/202307-31840/submission/359964/

## TODO
1. Add README.md
2. Add unittest

### CODE REVIEW

File: contrib/checkbox-provider-ce-oem/bin/serial_test.py [hunk 1]
```diff
@@ -50,92 +51,103 @@ def init_logger():
...
+            ser.rs485_mode = serial.rs485.RS485Settings()
+        ser.reset_input_buffer()
+        ser.reset_output_buffer()
+        return ser
+
+    def send(self, data: bytes) -> None:
+        try:
+            self.ser.rts = True
+            self.ser.write(data)
+            self.ser.rts = False
```
Comment (baconYao) [inline contrib/checkbox-provider-ce-oem/bin/serial_test.py:98]: Could you help me to understand why you switch the `rts` value before and after `write`?

File: contrib/checkbox-provider-ce-oem/bin/serial_test.py [hunk 2]
```diff
@@ -147,15 +159,28 @@ def client_mode(ser, data_length):
...
+    parser.add_argument(
+        "--type",
+        type=str,
+        help="The type of serial port (e.g. RS485, RS422, RS232, USB)",
+        default="USB"
+    )
+    parser.add_argument(
+        "--group",
+        type=str,
+        help="The group of serial ports that needed to be bringup also",
```
Comment (baconYao) [inline contrib/checkbox-provider-ce-oem/bin/serial_test.py:179]: `brinup` -> `brought up`

File: contrib/checkbox-provider-ce-oem/bin/serial_test.py [hunk 3]
```diff
@@ -50,92 +51,103 @@ def init_logger():
...
+            self.ser.rts = False
+            logging.info("Sent: {}".format(data.hex()))
+        except Exception:
+            logging.exception("Not able to send data!")
+
+    def recv(self) -> bytes:
+        rcv = ""
+        try:
+            self.ser.rts = False
+            rcv = self.ser.read(8)
```
Comment (baconYao) [inline contrib/checkbox-provider-ce-oem/bin/serial_test.py:107]: Why do you assign the `8` for reading?

File: contrib/checkbox-provider-ce-oem/bin/serial_test.py [hunk 4]
```diff
@@ -50,92 +51,103 @@ def init_logger():
...
+            if rcv:
+                logging.info("Received: {}".format(rcv.hex()))
+        except ValueError:
+            logging.exception("Received unmanageable string format")
+            rcv = "Error format"
```
Comment (baconYao) [inline contrib/checkbox-provider-ce-oem/bin/serial_test.py:112]: Why do you assign `Error format` here but never use it to do judgement?

**Review by baconYao — COMMENTED** [review]
Please see my inline comments

File: contrib/checkbox-provider-ce-oem/bin/serial_test.py [hunk 5]
```diff
@@ -50,92 +51,103 @@ def init_logger():
...
     the string send back from server. After receive the string,
     it will check the readback is correct or not.
     Usage:
     running on port /dev/ttymxc1 as a client
-    $ sudo ./rs485-remotr.py /dev/ttymxc1 --mode client
+    $ sudo ./serial_test.py /dev/ttymxc1 --mode client --type RS485
     """
-    test_str = "{}-{}".format(ser._port, str_generator(data_length))
-    sender(ser, test_str)
+    data = os.urandom(8)
```
[the remark this answers is not in the document]
Comment (stanley31huang) [reply contrib/checkbox-provider-ce-oem/bin/serial_test.py:143]: I think 8 bytes is too small, suggest we modify the data size to 1024 bytes

File: contrib/checkbox-provider-ce-oem/bin/serial_test.py [hunk 6]
```diff
@@ -50,92 +51,103 @@ def init_logger():
...
+    def __init__(
+        self,
+        node,
+        type,
+        group: list = [],
+        baudrate: int = 115200,
+        bytesize: int = 8,
+        parity: str = "N",
+        stopbits: int = 1,
+        timeout: int = 1,
```
[the remark this answers is not in the document]
Comment (LiaoU3) [reply contrib/checkbox-provider-ce-oem/bin/serial_test.py:64]: The default value for `serial.Serial` is 1, and it is usually 1.

File: contrib/checkbox-provider-ce-oem/bin/serial_console_test.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        ser.write("\n".encode())
+        response = ser.read(size=128).decode()
+        print("Received response:\n'{}'".format(response))
+        # ":~$" is the pattern for the DUT after logging in
+        # "login:" is the pattern for the DUT before logging in
+        if ":~$" in response or "login:" in response:
+            print("[PASS] Serial console test successful.")
+        else:
+            print("[FAIL] Serial console test failed.")
+            print("Expected response should contain ':~$' or 'login:'")
```
[the remark this answers is not in the document]
Comment (LiaoU3) [reply contrib/checkbox-provider-ce-oem/bin/serial_console_test.py:37]: Accident.

File: contrib/checkbox-provider-ce-oem/bin/serial_test.py (hunk 1 above)
Comment (LiaoU3) [reply contrib/checkbox-provider-ce-oem/bin/serial_test.py:98]: Actually we don't need this.

**Comment by LiaoU3** [comment]
@baconYao  Please review it when you have time.

File: contrib/checkbox-provider-ce-oem/bin/serial_console_test.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    try:
+        # Open serial port
+        with serial.Serial(serial_device, baudrate, timeout=1) as ser:
+            logger.info("Serial port opened successfully.")
+
+            # Send 'Enter Key'
+            logger.info("Sending 'Enter Key'...")
+            ser.write(os.linesep.encode())
+            response = ser.read(size=128).decode()
+            logger.info("Received response:\n'{}'".format(response))
```
Comment (stanley31huang) [inline contrib/checkbox-provider-ce-oem/bin/serial_console_test.py:31]: please use the correct way to log data with variable.
https://pylint.readthedocs.io/en/stable/user_guide/messages/warning/logging-fstring-interpolation.html

File: contrib/checkbox-provider-ce-oem/bin/serial_console_test.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        "-b",
+        default=115200,
+        help="The baud rate for the serial port",
+    )
+    args = parser.parse_args()
+    test_serial_console(args.device, args.baudrate)
+
+
+if __name__ == "__main__":
+    SystemExit(main())
```
Comment (stanley31huang) [inline contrib/checkbox-provider-ce-oem/bin/serial_console_test.py:67]: Do we really need to raise SystemExit here?

File: contrib/checkbox-provider-ce-oem/bin/serial_console_test.py (hunk 8 above)
Comment (LiaoU3) [reply contrib/checkbox-provider-ce-oem/bin/serial_console_test.py]: Done.

File: contrib/checkbox-provider-ce-oem/bin/serial_console_test.py (hunk 9 above)
Comment (LiaoU3) [reply contrib/checkbox-provider-ce-oem/bin/serial_console_test.py]: Done.

File: contrib/checkbox-provider-ce-oem/bin/serial_test.py [hunk 10]
```diff
@@ -50,92 +51,103 @@ def init_logger():
...
+class Serial:
+    def __init__(
+        self,
+        node,
+        type,
+        group: list = [],
+        baudrate: int = 115200,
+        bytesize: int = 8,
+        parity: str = "N",
+        stopbits: int = 1,
```
[the remark this answers is not in the document]
Comment (baconYao) [reply contrib/checkbox-provider-ce-oem/bin/serial_test.py:63]: Not only `parity` but `stopbits`, `bytesize` can be referenced.

File: contrib/checkbox-provider-ce-oem/bin/serial_console_test.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    )
+    parser.add_argument(
+        "--baudrate",
+        "-b",
+        default=115200,
+        help="The baud rate for the serial port",
+    )
+    args = parser.parse_args()
+    if len(args.device) != 1:
+        logger.error("Wrong port count. Should be 1.")
```
[the remark this answers is not in the document]
Comment (LiaoU3) [reply contrib/checkbox-provider-ce-oem/bin/serial_console_test.py]: I've removed it. It was an accident that it stays there.

**Review by baconYao — COMMENTED** [review]
See my inline comment

File: contrib/checkbox-provider-ce-oem/bin/serial_test.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (LiaoU3) [reply contrib/checkbox-provider-ce-oem/bin/serial_test.py]: Fixed

**Review by baconYao — COMMENTED** [review]
See my inline comment

File: contrib/checkbox-provider-ce-oem/bin/serial_test.py [hunk 12]
```diff
@@ -50,148 +51,193 @@ def init_logger():
...
         type=int,
-        help="set stopbits, one of {1, 2}, default: 1",
+        help="Stopbits",
+        default=1,
+    )
+    parser.add_argument(
+        "--datasize",
+        type=int,
+        help="Data size to send and receive",
+        default=128,
```
[the remark this answers is not in the document]
Comment (LiaoU3) [reply contrib/checkbox-provider-ce-oem/bin/serial_test.py:215]: 128

File: contrib/checkbox-provider-ce-oem/bin/serial_console_test.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+            # "login:" is the pattern for the DUT before logging in
+            if ":~$" in response or "login:" in response:
+                logger.info("[PASS] Serial console test successful.")
+            else:
+                logger.info("[FAIL] Serial console test failed.")
+                logger.info(
+                    "Expected response should contain ':~$' or 'login:'"
+                )
+                raise SystemExit(1)
+        logger.info("Serial port closeed successfully.")
```
Comment (baconYao) [inline contrib/checkbox-provider-ce-oem/bin/serial_console_test.py:42]: `closeed` -> `closed`

File: contrib/checkbox-provider-ce-oem/bin/serial_console_test.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+            ser.write(os.linesep.encode())
+            response = ser.read(size=128).decode()
+            logger.info("Received response: '%s'", response)
+            # ":~$" is the pattern for the DUT after logging in
+            # "login:" is the pattern for the DUT before logging in
+            if ":~$" in response or "login:" in response:
+                logger.info("[PASS] Serial console test successful.")
+            else:
+                logger.info("[FAIL] Serial console test failed.")
+                logger.info(
```
Comment (baconYao) [inline contrib/checkbox-provider-ce-oem/bin/serial_console_test.py:38]: `logger.info` -> `logger.error`

**Review by baconYao — COMMENTED** [review]
See my inline comment

File: contrib/checkbox-provider-ce-oem/units/serial/jobs.pxu [hunk 15]
```diff
@@ -1,36 +1,95 @@
...
+    pairs=($SERIAL_CONSOLE_PORTS)
+    for pair in "${pairs[@]}"; do
+        IFS=':' read -r -a pair_array <<< "$pair"
+        type="${pair_array[0]}"
+        node="${pair_array[1]}"
+        baudrate="${pair_array[2]}"
+        echo "type: $type"
+        echo "node: $node"
+        echo "baudrate: $baudrate"
+        echo ""
```
Comment (rickwu666666) [inline contrib/checkbox-provider-ce-oem/units/serial/jobs.pxu:27]: This part should be able to leverage "serial_config_parser.py".

**Review by rickwu666666 — DISMISSED** [review]
Please see my in-line comment.

File: contrib/checkbox-provider-ce-oem/bin/serial_console_test.py (hunk 14 above)
Comment (LiaoU3) [reply contrib/checkbox-provider-ce-oem/bin/serial_console_test.py:38]: I think `error` is for the time when code is not running in the right way such as any unexpected behavior. In this case, program run without error, so it should be `info`
What do you say?

File: contrib/checkbox-provider-ce-oem/units/serial/jobs.pxu [hunk 16]
```diff
@@ -1,36 +1,95 @@
...
+
+unit: template
+template-resource: ce-oem-serial/serial-list
+template-unit: job
+template-engine: jinja2
+template-id: ce-oem-serial/serial-transmit-data-tests
+id: ce-oem-serial/serial-transmit-data-{{ type }}-{{ node }}-{{ baudrate }}
+# imports: from com.canonical.plainbox import manifest
+# requires: 
+#     manifest.has_serial_ehco_server == True
```
[the remark this answers is not in the document]
Comment (LiaoU3) [reply contrib/checkbox-provider-ce-oem/units/serial/jobs.pxu]: Fixed.

File: contrib/checkbox-provider-ce-oem/units/serial/jobs.pxu [hunk 17]
```diff
@@ -1,36 +1,95 @@
...
-_summary: To check the port {RS485} can working on RS485 half-duplex mode.
-_purpose:
-    To check the port {RS485} can send and receive the string with
-    RS485 half-duplex mode.
+template-engine: jinja2
+template-id: ce-oem-serial/serial-console-tests
+id: ce-oem-serial/serial-console-{{ type }}-{{ node }}-{{ baudrate }}
+# imports: from com.canonical.plainbox import manifest
+# requires: 
+#     manifest.has_serial_console_loopback == True
```
[the remark this answers is not in the document]
Comment (LiaoU3) [reply contrib/checkbox-provider-ce-oem/units/serial/jobs.pxu]: Fixed.

**Review by baconYao — DISMISSED** [review]
LGTM

**Review by stanley31huang — APPROVED** [review]
LGTM

**Review by baconYao — APPROVED** [review]
+1

**Pull request merged**